src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -185,19 +188,31 @@ protected SCMRevision retrieve(@NonNull SCMHead head, @NonNull TaskListener list
             listener.getLogger().println("Fetching " + remoteName + "...");
             List<RefSpec> refSpecs = getRefSpecs();
             client.fetch(remoteName, refSpecs.toArray(new RefSpec[refSpecs.size()]));
-            // we don't prune remotes here, as we just want one head's revision
-            for (Branch b : client.getRemoteBranches()) {
-                String branchName = StringUtils.removeStart(b.getName(), remoteName + "/");
-                if (branchName.equals(head.getName())) {
-                    return new SCMRevisionImpl(head, b.getSHA1String());
-                }
-            }
-            return null;
+            return retriever.run(client, remoteName);
```

Comment:
Big diff blocks are just removing duplication present in @stephenc’s original implementation, preparing to use `doRetrieve` some more times.